kasa/protocol.py

```diff
@@ -25,23 +25,102 @@
 _NO_RETRY_ERRORS = {errno.EHOSTDOWN, errno.EHOSTUNREACH, errno.ECONNREFUSED}
 
 
-class TPLinkSmartHomeProtocol:
+class TPLinkProtocol:
+    """Base class for all TP-Link Smart Home communication."""
+
+    DEFAULT_TIMEOUT = 5
+
+    def __init__(self, host: str, *, port: Optional[int] = None) -> None:
+        """Create a protocol object."""
+        self.host = host
+        self.port = port
+        self.timeout = self.DEFAULT_TIMEOUT
+
+    @staticmethod
+    def get_discovery_targets(targetip: str = "255.255.255.255"):
+        """Return a list of discovery targets for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("get_discovery_targets should be overridden")
+
+    @staticmethod
+    def get_discovery_payload():
+        """Get discovery payload for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("get_discovery_payload should be overridden")
+
+    @staticmethod
+    def try_get_discovery_info(port, data):
+        """Try get the discovery info from supplied data after a discovery broadcast has resulted in a received datagram."""
+        raise SmartDeviceException("try_get_discovery_info should be overridden")
+
+    async def try_query_discovery_info(self):
+        """Try to query a device for discovery info."""
+        raise SmartDeviceException("try_query_discovery_info should be overridden")
+
+    async def query(self, request: Union[str, Dict], retry_count: int = 3) -> Dict:
+        """Query the device for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("query should be overridden")
+
+    async def close(self) -> None:
+        """Close the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("close should be overridden")
```

Comment:
This can be `abs.abstractmethod` instead

Reply:
Will do